array.h

```diff
@@ -908,6 +917,43 @@ template <size_t Rank, size_t... Is>
 using enable_if_permutation = std::enable_if_t<sizeof...(Is) == Rank && all(Is < Rank...) &&
                                                product((Is + 2)...) == factorial(Rank + 1)>;
 
+template <index_t DimIndex, index_t Min, index_t Extent, index_t Stride,
+          class DimSrc>
+NDARRAY_HOST_DEVICE void assert_dim_compatible(const dim<Min, Extent, Stride>&, const DimSrc& src) {
+  bool compatible = true;
+  if (is_static(Min) && src.min() != Min) {
+    NDARRAY_PRINT_ERR("Error converting dim %zu: expected static min " NDARRAY_INDEX_T_FMT ", got " NDARRAY_INDEX_T_FMT "\n",
+            DimIndex, Min, src.min());
+    compatible = false;
+  }
+  if (is_static(Extent) && src.extent() != Extent) {
+    NDARRAY_PRINT_ERR("Error converting dim %zu: expected static extent " NDARRAY_INDEX_T_FMT ", got " NDARRAY_INDEX_T_FMT "\n",
+            DimIndex, Extent, src.extent());
+    compatible = false;
+  }
+  if (is_static(Stride) && src.stride() != Stride) {
+    NDARRAY_PRINT_ERR("Error converting dim %zu: expected static stride " NDARRAY_INDEX_T_FMT ", got " NDARRAY_INDEX_T_FMT "\n",
+            DimIndex, Stride, src.stride());
+    compatible = false;
+  }
+  assert(compatible);
+  (void) compatible;
+}
+
+template <class DimsDst, class DimsSrc, size_t... Is>
+NDARRAY_HOST_DEVICE void assert_dims_compatible(const DimsDst&, const DimsSrc& src, index_sequence<Is...>) {
+  // This is ugly, in C++17, we'd use a fold  expression over the comma operator (f(), ...).
+  int unused[] = {(assert_dim_compatible<Is>(typename std::tuple_element<Is, DimsDst>::type(), std::get<Is>(src)), 0)...};
+  (void) unused;
+}
+
+template <class...DimsDst, class DimsSrc>
+NDARRAY_HOST_DEVICE void assert_dims_compatible(const DimsSrc& src) {
+#ifndef NDEBUG
+  assert_dims_compatible(DimsDst()..., src, std::make_index_sequence<std::tuple_size<DimsDst...>::value>());
```

Comment:
Just `make_index_sequence`, it's pulled into the internal namespace. At one point I was just providing my own implementation of it to avoid a dependency on a C++ version, but now it is just for consistency.

Reply:
Done.